### PR TITLE
Use `Base._unsetindex!` in `pop!` and `popfirst!`

### LINKED ISSUE #884: Deque leaks memory
Deque uses `data::Vector{T}` in `DequeBlock{T}`, but it only empties this vector in rear blocks, never touching this vector in the head block, instead it only invokes `reset!` on it. But `reset!` doesn't unset index on data elements that are no longer valid in this head block, causing these objects not to be garbage collected.

I verified this by taking heap snapshot after inserting and dequeuing all elements and running `GC.gc(true)` - the elements were still in the heap.

Not sure if other data structures have this behaviour, I only checked for Deque (and by proxy Queue)

### PR DESCRIPTION
**Pull request by kpamnany**
For `Deque` and `CircularDeque`. So that popped elements are not rooted by the deque and can be GCed when they drop out of caller scope.

Completes the fix of #884.

### CODE REVIEW

File: src/deque.jl
Comment (nickrobinson251) [inline src/deque.jl:326]: i wonder if these lines are worthy of a comment? maybe even just 
```suggestion
    Base._unsetindex!(head.data, head.front) # see issue/884
```

File: src/deque.jl
Comment (nickrobinson251) [inline src/deque.jl:300]:
```suggestion
    Base._unsetindex!(rear.data, rear.back) # see issue/884
```

**Review by nickrobinson251 — APPROVED** [review]
Thanks, @kpamnany ! LGTM

@oxinabox are you still somewhat of a maintainer here? It'd be very helpful for us (at RAI) to get a new release with this fix 🙏  Or perhaps you could give me permissions on this repo? (I used to have them but never asked for them back when i lost all perms because of a MFA change) -- thanks, Frames!

**Comment by oxinabox** [comment]
@nickrobinson251   I am still maintaining things here, but I am happy to hand this one over to you.
Go to https://github.com/orgs/JuliaCollections/ to accept your invite back to the org.

Usual procedure once happy with it: merge it into master, and backport it into `#0.18-release`,

**Comment by nickrobinson251** [comment]
Also i wonder if we could add a test for this? Something based on checking `summarysize` perhaps?

File: src/circ_deque.jl
Comment (kpamnany) [inline src/circ_deque.jl:67]:
```suggestion
    Base._unsetindex!(D.buffer, D.last) # see issue/884
```

File: src/circ_deque.jl
Comment (kpamnany) [inline src/circ_deque.jl:95]:
```suggestion
    Base._unsetindex!(D.buffer, D.first) # see issue/884
```

File: src/circ_deque.jl
Comment (nickrobinson251) [inline src/circ_deque.jl:67]:
```suggestion
    Base._unsetindex!(D.buffer, D.last)
```

**Pull request merged**
